### ISSUE TITLE
AWS CW API Throttle Issue

### ISSUE THREAD

**Issue by mr337** [aws-robotics/cloudwatchlogs-ros1#21]
Hi,

I have been running into an issue with hitting the API limits and throttled back. I thought the issue was related to https://github.com/aws-robotics/cloudwatchlogs-ros1/issues/20 generating a lot of log entries, but that has been fixed and confirmed.

I am still getting the dreaded ThrottlingException. I have tried various `publish_frequency`

`publish_frequency` = 10s
```
[ERROR] [1554434092.587344572, 32.603000000]: [AWSClient] HTTP response code: 400
Exception name: ThrottlingException
Error message: Rate exceeded
5 response headers:
connection : close
content-length : 58
content-type : application/x-amz-json-1.1
date : Fri, 05 Apr 2019 03:14:51 GMT
x-amzn-requestid : faa1d28b-5750-11e9-9b5c-49f40dc0bbd4
```

`publish_frequency` = 60s
```
[ERROR] [1554433884.447696369, 406.472000000]: [AWSClient] HTTP response code: 400
Exception name: ThrottlingException
Error message: Rate exceeded
5 response headers:
connection : close
content-length : 58
content-type : application/x-amz-json-1.1
date : Fri, 05 Apr 2019 03:11:23 GMT
x-amzn-requestid : 7e92c988-5750-11e9-a09f-0db3a37537c2
```

When this is encountered from what I can tell the batch of logs sent go missing. Majority of the time logging will resume and show in CW and other times it just continuously error out.

**Comment by AAlon** [comment]
Roughly, how many lines are written to rosout and at what frequency? 
We'll need to improve the aggregation and batching of logs in order to make fewer API calls. This feedback is extremely important to us - thank you! 

To unblock you it might be quicker to pursue a limit increase for the PutLogEvents API, but it depends on how high the call volume is relative to the default limit. We can look into this for you if you'd like; you can email the relevant AWS account ID to ros-contributions@amazon.com and refer to this issue.

**Comment by mr337** [comment]
I will have to get exact number but normally 1-2 lines/second with spikes up to 10 lines/second. At the moment we are logging anything DEBUG level for testing and ran into the issue.

I'll send an email out so we can get exact numbers.

**Comment by nburek** [comment]
It would actually be a little surprising to me if we are throttling based on the number of logs being published. 

I've been able to use `rostopic pub` to send 100 messages per second with the default 5 second batching rate without issue. With this configuration we would be publishing 500 logs at once every 5 seconds. From what you've said this should be more than your system is logging. 

The `PutLogEvents` api that we use to post logs has a default throttle rate of 5 calls per second per stream. If you are running multiple instances of the the logger node (even across different systems) make sure that they are using different log streams. Our recommendation is to have every robot you have running the application use the same log group but have a unique log stream per robot. 

To me the more likely culprit is the fact that we are calling the DescribeLogStreams api every time the publisher runs in order to get the stream token. This API is throttled at a rate of 5 calls per second per account per region. This means that the likelihood of exceeding that call rate increases with the number of instances you have running and and as the `publiser_frequency` lowers (making it more likely that the multiple instances will be synchronized to send within the same second). 

I will work on the fix for not calling DescribeLogStreams as often. @mr337 Can you confirm if the scenario I described above sounds like your setup?

**Comment by nburek** [comment]
Another note is that we limit every call to PutLogEvents to 100 log events and that we don't have any client side throttling on this. This means that if you're generating more than 500 log message lines within the publish_freqency rate you have set you may see Throttled calls on the PutLogEvents call.

**Comment by mr337** [comment]
Hi @nburek

> I will work on the fix for not calling DescribeLogStreams as often. @mr337 Can you confirm if the scenario I described above sounds like your setup?

Sadly that is not the case. The testing and results in the original post was just evaluating the logging node. It was in simulation with a single instance running, no multiple instances or heavy messaging. Everything was writing to a single log group/stream. I barely generated any traffic to hit a threshold...confused :)

>  Another note is that we limit every call to PutLogEvents to 100 log events and that we don't have any client side throttling on this. This means that if you're generating more than 500 log message lines within the publish_freqency rate you have set you may see Throttled calls on the PutLogEvents call.

That is nice to know and I'll keep that in mind. I might have hit that when testing very long `publish_frequency` of several minutes creating a big batch.

@AAlon I have sent in an email referring to this issue.

If either of you need any follow up testing let me know!

**Comment by xabxx** [comment]
@mr337 hi there. We've pushed a fix for this issue. Did you have a chance to try it yet?

**Comment by nburek** [comment]
@xabxx @mr337 To be clear, the fix has not been pushed to mainline, but is in the `get_token_fix` branch of the `cloudwatch_common` repository with a pull request ongoing to get it merged into the master branch there.

**Comment by mr337** [comment]
Appreciate the update! I haven't applied the fix (@nburek provided) or tried the `get_toekn_fix` branch yet. Plan on circling back shortly to give it a try.

**Comment by xabxx** [comment]
@mr337 thank you! Please keep us updated! :)

**Comment by mr337** [comment]
So pulled down both cloudwatch-ros1 and cloudwatch-common and compiled.

Here are the checkouts of both repos
```
/tmp/ros-workspace/src/cloudwatch-common (get_token_fix) $ git branch -v
* get_token_fix ca10e4f Fixes issue with DescribeLogStreams being called to get the next token when it isn't necessary.

/tmp/ros-workspace/src/cloudwatchlogs-ros1 (release-v2) $ git branch -v
* release-v2 6e5b343 2.0.0
```

I am unclear if I need to use `cloudwatchlog-ros1 release-v2`. Using `release-1` I had compile issues. 

After running for a few hours with regular logging I have no issues with logs correctly being pushed. Every now and then I will the following which is filtered out and doesn't seem like anything bad.
```
[ INFO] [1556254136.453533847, 1530.411000000]: [EnvironmentAWSCredentialsProvider] Found secret key
[ INFO] [1556254136.453803877, 1530.411000000]: [CurlHandleContainer] Connection has been released. Continuing.

```

**Comment by juanrh** [comment]
Hi @mr337 mr337, thanks a lot for the update. You used the correct branch, in general the latest release branch should be used, which in this case is `release-v2` for `cloudwatchlogs-ros1`. 

Those messages are related to the AWS SDK accessing to the credentials to connect to the CloudWatch logs service, and connecting to the service, so they are expected (see [EnvironmentAWSCredentialsProvider](https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/include/aws/core/auth/AWSCredentialsProvider.h#L126) and [CurlHandleContainer](https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/include/aws/core/http/curl/CurlHandleContainer.h)).

I'm closing this issue, but please feel free to re-open if you have any other problems. We'll also update this ticket when that PR is merged to master

**Comment by juanrh** [comment]
reopening ticket, we'll keep it open until we release the code fix in Apt

**Comment by timrobotson** [comment]
Closing this as it's fixed in code. We're switching to a style similar to most open source projects where we'll close PR's when the code is fixed, then when we do a release we'll make a new PR and link back to this PR so they'll be a notification here when it's done.
